RangeControls: Improve UI + UX

ItsJonQ opened the pull request:
## Description
<img width="661" alt="Screen Shot 2020-01-27 at 11 42 09 AM" src="https://user-images.githubusercontent.com/2322354/73198465-069dd080-4101-11ea-9292-78477d39f2ed.png">

This update improves the UI and UX for the `RangeControl` component with `@wordpress/components`.

### Features ✨
* Visual tooltip that displays current value (Tooltip). Follows the "thumb" (circle)
* **Larger hitbox**, making it easier to click/drag
* More consistent UI between slider, input, and Reset button
* Renders step markers with a simple + custom prop (marks)
* Tooltip auto positions top or bottom, based on ideal rendering position
* Can easily change the color of the "track" using a color prop
* Can customize the tooltip's content with a custom renderProp
* Ability to show/hide the accompanying `input[type="number"]`

Example of step marks + Tooltip rendering:

![](https://user-images.githubusercontent.com/2322354/73027347-3bf4a680-3e01-11ea-83b0-a144f12b3fc0.gif)

### A11Y Considerations 💪
* Uses native `input[range]` under the hood, which gives us native keyboard interaction handling
* Uses `aria` roles on elements to emphasize changes / reduce noise (via `aria-hidden`)
* Respects reduce-motion to disable all animations


### Other Considerations ✌️

<img width="792" alt="Screen Shot 2020-01-27 at 12 25 03 PM" src="https://user-images.githubusercontent.com/2322354/73199090-34374980-4102-11ea-81bd-ffaf577654b4.png">

* RTL support
* README file has been updated with additive API changes


## How has this been tested?

![](https://user-images.githubusercontent.com/2322354/73111050-dbd03400-3ed6-11ea-8f10-a1f2070c59ef.gif)

This update was developed within Storybook. When ready, it was integrated and tested with Gutenberg, local dev.

### Voiceover
This component was tested **with Mac OS Voiceover**, to ensure the value change announcements + keyboard interactions worked as expected.

### Mobile

![IMAGE 2020-01-27 11:35:06](https://user-images.githubusercontent.com/2322354/73199567-10c0ce80-4103-11ea-9cb9-fd9214e57fe9.jpg)

Although, this component only updates the **Web** version, the changes were tested on both iOS (iPhone 8) and Android (Samsung A50)

[iPhone 8 Screen Recording](https://d.pr/v/gH5nKS)

Note: The `inputMode` for the number input was set to `decimal`, which allows for the `numpad` keyboard to appear for mobile.

## Types of changes

### CSS

The CSS for the updated component was written with Emotion, leveraging the new/experimental CSS-in-JS system.

### Tooltip

The Tooltip that appears to indicate the value is **not** a `@wordpress/components` one, but rather, a simple/custom implementation.

**Reason**

I had originally tried using the Popover/Tooltip from `@wordpress/components`, but this led to (perceived) performance rendering issues:

https://d.pr/i/a6i3R4

The positioning render cycle was not in sync with the quick changes, which made the experience look and feel really laggy.

As a work-around, I created a simple Tooltip which is currently only being used in the new `RangeControl` component. (I'm open to changing this approach if anyone has suggestions)


### Backwards Compatibility
This update preserves all of the Component APIs. Therefore, it shouldn't break backwards compatibility. It passes all of the previous unit tests and appears to be working as expected when testing within Gutenberg. Component API changes are **additive**.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

Resolves: https://github.com/WordPress/gutenberg/issues/19845

Commit:
Start with new RangeControl component

Commit:
Add tooltipPosition to RangeControl

Commit:
Improve styles and interactions for new RangeControl

Commit:
Add custom marks support

Commit:
Updated UI for Ranger Control

Commit:
Improve layout rendering of RangeControl

Commit:
Add wrapper for legacy/next RangeControl.

Also test integration with Gutenberg

Commit:
Adjusting range-control styles

Commit:
Update RangeControl UI and snapshots

Commit:
Darken thumb color

Commit:
Consolidate legacy/next implementation. Update styles

Commit:
Update snapshot + remove RangeControlNext test

Commit:
Update README for RangeControl

Commit:
Add RTL support

Commit:
Update storybook snapshots

Commit:
Remove unused Range Control SCSS

ZebulanStanphill commented on packages/components/src/range-control/README.md:
```diff
@@ -152,6 +152,22 @@ If this property is true, a button to reset the the slider is rendered.
 - Required: No
 - Platform: Web | Mobile
 
+#### color
```
What's the use cases for changing the color of the filled range? Wouldn't you just use CSS for this sort of thing?

ItsJonQ replied:
Halloo 👋 @ZebulanStanphill !

That's a good question :). For the components rendering Gutenberg/app UI (the `@wordpress/components` package), I believe customization should be done based on what a component exposes (it's API). 

Sort of like how a `<Button />`'s size can be adjusted with an `isSmall`, rather than using CSS for `height` and so on.

The UI can certainly be adjusted with CSS (any UI could really). But doing so is brittle, as selectors (usually classNames) are subject to change in the future.

In the case of the updated `RangeControl`, I've exposed the (track) colour, which can better accommodate upcoming UI updates (from the G2 initiative), as well as custom block developers.

That being said, we don't have to go with this implementation

Would love to hear what others think :)

youknowriad replied:
While I agree that we need a consistent way to change the color for different controls, this PR is the first that tries to introduce one. 

Options here could be a "theme" prop as well instead of very specific design props. 

So while I'm not against it in principle, I wonder if we should avoid adding these props for now and think about this kind of customization more globally?

ItsJonQ replied:
@youknowriad Oo! A `theme` prop sounds like a good API to encapsulate design props.

> I wonder if we should avoid adding these props for now and think about this kind of customization more globally?

I can omit the addition of this property for now. A global system for handling theming would be lovely, and it's something that [Emotion Theming](https://emotion.sh/docs/theming) supports. 

For instances where you may want to customize the aesthetic of a very specific, the `ThemeProvider` can be added to handle that.

ItsJonQ replied:
@youknowriad Update! For now, I've removed the `color` prop. We can approach theming in another PR.

karmatosed commented:
This is great!  A few pieces of feedback to talk about and then happy to review.

- Could we have no shadow perhaps as I'm not sure we had before on the circle?
- Are we able to increase the size of the arrows on input, I would love that to be bigger.

Commit:
Remove color prop and shadow. Improve focus interaction

Also add comments for hooks and functions

ItsJonQ commented:
@karmatosed Haii!

> Could we have no shadow perhaps as I'm not sure we had before on the circle?
Shadow removed. In doing so, I noticed that there was barely any visual feedback to indicate a focus state. I've adjusted it so that the circle has a blue ring (similar to `input` fields) when focused:

<img width="588" alt="Screen Shot 2020-01-29 at 9 06 58 AM" src="https://user-images.githubusercontent.com/2322354/73364812-366fe400-4279-11ea-9073-68e107a5114b.png">

> Are we able to increase the size of the arrows on input, I would love that to be bigger.
That `input` field is currently using an out of the box number `input`. How they render is based on your browser x OS.

To make the arrows bigger, we'd need to do quite a bit of custom UI/CSS work. In which case, I'd advise perhaps updating the base `TextControl` to support this (in a separate PR).

Hope that makes sense!

karmatosed commented:
> To make the arrows bigger, we'd need to do quite a bit of custom UI/CSS work. In which case, I'd advise perhaps updating the base TextControl to support this (in a separate PR).

That makes sense, I would love to get that happening if we can as I think it could benefit us and avoid those tiny arrows.

This all looks great and let's look at getting this in.

karmatosed reviewed (approved):
Approving design with the note the arrows needs iterating outside this.

youknowriad commented on packages/components/src/range-control/tooltip.js:
```diff
@@ -0,0 +1,82 @@
+
+const TOOLTIP_OFFSET_HEIGHT = 32;
+
+export default function SimpleTooltip( {
```
We already have a Tooltip component, what's preventing us from using it instead?

ItsJonQ replied:
@youknowriad Haii! That's a good question :). I explained it a bit in the description. I'm happy to expand on it further if it helps.

Essentially, the positioning mechanism of the `Tooltip` component (aka. `Popover` under the hood) is visually laggy, as seen here:

https://d.pr/i/a6i3R4

(Note: in that demo, I used `Popover`, but the mechanism is the same)

Because of that, I opted to create a simple Tooltip just for this component. Granted, I don't think that's the ideal solution long term.

Hope this helps!

Commit:
Formatted RangeControl files

Commit:
Merge branch 'master' into try/range-controls-ui-update

Commit:
Update storybook snapshot

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -150,28 +151,30 @@ export const ThumbWrapper = styled.span`
 	user-select: none;
 
-	[dir=rtl] & {
+	[dir='rtl'] & {
```
is it possible to have bable plugin or something that generates the RTL styles like we do with sass?

ItsJonQ replied:
@youknowriad Not at this very moment, although having an automated solution would be lovely!

CSS-in-JS libraries need some way to adjust the CSS compiler output (`stylis`), typically with a Provider of some kind, ([discussion + examples](https://github.com/emotion-js/emotion/issues/793))

Commit:
Format rail.js file

youknowriad commented on packages/components/src/range-control/README.md:
```diff
@@ -193,6 +229,86 @@ The maximum value accepted. If higher values are inserted onChange will not be c
+
+- Type: `Number`
+- Required: No
+- Platform: Web
```
Do we really need all these customization options: `tooltiipPosition`, `tooltipTimetout` and `tooltipZIndex`? I think generally speaking for regular design systems, they tend to be very liberal to add props and variations but in the context of @wordpress/components we prefer the opposite because the backward compatibility strategy is stricter than what npm packages do in general.

Maybe we can just provide sane defaults and introduce props if we really get important use-cases for them?

ItsJonQ replied:
> Maybe we can just provide sane defaults and introduce props if we really get important use-cases for them?

Can do. They already have default values. I'll remove the public facing APIs for these.

youknowriad commented on packages/components/src/range-control/README.md:
```diff
@@ -193,6 +229,86 @@ The maximum value accepted. If higher values are inserted onChange will not be c
+- Required: Yes
+- Platform: Web | Mobile
+
+#### width
```
Do we need the width prop or can the component be like 100% by default?

ItsJonQ replied:
Ah! Good catch. This prop isn't required. Will make that change.

Commit:
Reduce component public API to adjust aesthetics / interactions

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+
+export const Wrapper = styled.span`
+	box-sizing: border-box;
+	color: ${color( 'blue.medium.focus' )};
```
I see several cases of "blue" usage in this file. Do these adapt to the WordPress admin scheme colors? Should they?
we have the `theme(something)` colors in SCSS to solve that problem there.

ItsJonQ replied:
@youknowriad These colours correspond to the [(SCSS) based colour values](https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_colors.scss), but are retrieved via the `color()` util function. The original mappings are here:

https://github.com/WordPress/gutenberg/blob/master/packages/components/src/utils/colors-values.js#L76

youknowriad replied:
but what is the expectation here, do we want these colors to adapt to WordPress's admin scheme colors or not?

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+`;
+
+export const BeforeIconWrapper = styled.span`
+	margin-right: 6px;
```
Do we need an RTL style for this?

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+`;
+
+export const AfterIconWrapper = styled.span`
+	margin-left: 16px;
```
Do we need an RTL style for this?

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+	margin-right: 6px;
+	margin-top: 5px;
+
+	> .dashicon {
```
Where are we using Dashicons here? can we replace them with an icon from the icons package?

ItsJonQ replied:
@youknowriad I preserved the same functionality as the previous version, which uses Dashicon.

I just pushed an update that removes `Dashicon` usage and opts for the new `@wordpress/icon` instead. (Tests have also been updated)

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+	}
+`;
+
+const handleReducedMotion = () => {
```
Feels like something that could be moved to the utils folder (like color) and should we align its implementation with the SASS version we have?

ItsJonQ replied:
Ah, good idea 👍

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,310 @@
+export const ActionRightWrapper = styled.span`
+	box-sizing: border-box;
+	display: block;
+	margin-left: 8px;
```
Each time I see margin-left (or right or padding), I wonder whether we need RTL equivalents. Maybe we can solve these with mixins for now, otherwise, we'll quickly forget about these.

youknowriad replied:
What about this one, does it need to be wrapped in rtl?

Commit:
Add RTL knob for RangeControl story

Commit:
Update Storybook snapshot

Commit:
Add RTL renderer (cssjanus) and style mixins

Commit:
Replace Dashicon with @wordpress/icons. Update RTL usage in Component

Commit:
Update rtl rendering in RangeControl styles

Fix tests to check for svg via @wordpress/icon vs Dashicon

youknowriad commented on packages/components/src/range-control/index.js:
```diff
@@ -15,14 +15,14 @@ import {
 	useState,
 	forwardRef,
 } from '@wordpress/element';
+import * as WPIcons from '@wordpress/icons';
```
`import *` should be avoided for the icons package because it breaks tree-shaking causing us to bundle the whole script here. We should only pick the icons we're using.

ItsJonQ commented:
@youknowriad Thanks for your thorough review!! Really appreciate it 😊 

I've made several updates regarding your concerns.

### Reduce motion mixin

I've abstracted the function into a new `reduceMotion` mixin that works just like the SCSS counterpart. There is now a file (`style-mixins.js`) that exports it, along with other utils for CSS-in-JS styling.

### Better RTL style handling

There isn't seamless RTL rendering for CSS-in-JS yet. However, all of the proposed techniques use the same transformation engine under the hood, [CSS Janus](https://github.com/cssjanus/cssjanus) ([4.3kb, 1.6kb gzipped](https://bundlephobia.com/result?p=cssjanus@1.3.2)). I've added it to `@wordpress/components`.

To use it, I've created an `rtl` mixin that can be used in CSS-in-JS for transformations.
It isn't zero overhead, but it's much better than my previous implementation (thanks again for pointing it out).

Let me know what you think

### ~Dashicon~ 👉 `@wordpress/icons`

I kept `Dashicon` for the original PR, as that was the previous implementation. I know there's been a lot of icon work (great job BTW!), so it makes sense for us to make that update. Based on your suggestion, I've updated the code to use `@wordpress/icons` and removed `Dashicon` entirely from `RangeControl`

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,288 @@
+	touch-action: none;
+	width: 100%;
+
+	${rtl( {
```
I think, I'm not sure that the `rtl` mixin is based on the direction of the document. Whicih makes me wonder whether this style here is needed since the direction is already applied above?

ItsJonQ replied:
Ah, you are correct! I had thought I needed it during early development of this. I just tried removing it, and it works as expected 👍

Commit:
Fix styling for ActionRightWrapper

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,288 @@
+
+export const ThumbWrapper = styled.span`
+	align-items: center;
+	margin-left: -10px;
```
Should we remove this one? (as it's added with rtl support down)

youknowriad commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,288 @@
+		height: 0;
+		left: 50%;
+		line-height: 0;
+		margin-left: -6px;
```
do we need rtl for this,

ItsJonQ replied:
Not needed for this one. The combination of `left: 50%` with `margin-left: -6px` is used for centering. Filling them to `right` and `margin-right` would yield the same result.

youknowriad reviewed (commented):
Thanks for the update @ItsJonQ this is looking very good for me.

I admit that I'm still new to CSS in JS so still trying to find my way there. I think the "rtl" mixin added is something good that contributors should be aware of to avoid forgetting about RTL (I wonder how we can force it, lint it later). I like the reduceMotion mixin too. I'd like another opinion on the changes here. cc @aduth @gziolo 

- The only blocker for me is the `import *`

And Ideally, i'd like us to follow-up and avoid the adhoc Tooltip component maybe by improving the existing one.

ItsJonQ commented:
@youknowriad Thank you 🙏 !

I'm not sure the best way to proceed with the Icons.

Previously, you could pass a string, and it would render the desired Icon (`Dashicon`).
However, this isn't the case anymore if we want to adhere to the new `<Icon icon={iconComponent} />` API.

I did the wildcard `*` import to preserve that functionality (passing `string` into `beforeIcon` and `afterIcon`). However, you make a very valid point in regards to it nullify'ing the benefits of treeshaking

youknowriad commented:
> Previously, you could pass a string, and it would render the desired Icon (Dashicon).

Oh right! I guess we can keep the previous behavior for icons for now until we figure out a sunset plan for Dashicon (See #20003 )

youknowriad commented:
<img width="274" alt="Capture d’écran 2020-02-06 à 6 14 04 PM" src="https://user-images.githubusercontent.com/272444/73961328-a7b03680-490c-11ea-8c23-fae905e33741.png">

I noticed a small visual breakage in the font size selector of the paragraph block (the custom input is large). I'm not sure it's entirely related to this branch though.

Commit:
Remove unnecessary RTL. Adjust RTL in ThumbWrapper

Commit:
Restore Dashicon with RangeControl

Restoring Dashicon with RangeControl for now, until we can figure out a
better sunsetting plan

ItsJonQ commented:
> Edit I confirm, it's specific to this branch.
@youknowriad 😱 ! I see it too. Investigating 💪

Commit:
Add style fix for RangeControl x FontSizePicker

ItsJonQ commented:
@youknowriad I just pushed my fix. The issue was the `.components-range-control__number` className was being used ad-hoc outside of `RangeControl`, specifically in `FontSizePicker`.

I've fixed it by adding the style rule back in, with a comment:

```scss
// Keeping this className style rule as it is currently being
// used OUTSIDE of the RangeControl component.
.components-range-control__number {
	display: inline-block;
	margin-left: $grid-size;
	font-weight: 500;
	width: 54px;
}
```

youknowriad commented on packages/components/src/range-control/style.scss:
```diff
@@ -0,0 +1,8 @@
+// Keeping this className style rule as it is currently being
+// used OUTSIDE of the RangeControl component.
```
Bad :)

youknowriad replied:
I did a quick search and I noticed that it's only used in FontSizePicker. 
Maybe we can just move the style there and rename the class to follow our guidelines `components-font-size-picker__number`

ItsJonQ replied:
Sounds good to me! I'll do that now

youknowriad reviewed (approved):
LGTM 👍 Thanks for your hard work here.

Commit:
Migrate on RangeControl styles to FontSizePicker

Rename rangeControl related ID and classNames to match the structure
of FontSizePicker

ItsJonQ commented:
@youknowriad 

> Module cssjanus has an incompatible license 'Apache-2.0'

Aww. Does that mean we can't use this package?

ZebulanStanphill commented:
Just to clarify, Apache 2.0 is compatible with GPLv3, but not earlier versions of the GPL. WordPress is GPLv2+ (which means it could legally switch to GPLv3+ to use this library, but dropping GPLv2 compatibility is probably not a reasonable option here).

If you really want to use this particular library, you could ask the library contributors (there are 12 of them) to consider relicensing or dual-licensing under a license compatible with GPLv2+.

ItsJonQ commented:
@ZebulanStanphill Thanks for your insight! I'm going to explore alternative solutions now

Commit:
Replace cssjanus with basic rtl solution

ItsJonQ commented:
Updates! I took a look at various solutions, and nothing seemed to feel right. I've opted to remove cssjanus, and wrote a tiny albeit naively simple and basic utility.

cc'ing @youknowriad and @aduth for thoughts!

(Sidenote: My Regex game is quite weak 😂 . Please feel free to improve the implementation!)

It's definitely not as seamless as `cssjanus` or the `postcss` auto-generated workflow.
I can see this working short term... but would definitely love something better for the long term.

youknowriad commented:
Looks like there's some valid e2e test failures.

ItsJonQ commented:
@youknowriad Ah. Looks like I need to update the E2E selectors for `FontSizePicker`, since we renamed them

Commit:
Fix e2e test for FontSizePicker

By updating the className for the input number field

ItsJonQ commented:
Yay! E2E is fixed 😍

jsnajdr commented on packages/components/src/range-control/styles/range-control-styles.js:
```diff
@@ -0,0 +1,287 @@
+	top: -4px;
+	width: 1px;
+
+	${markFill};
```
It's a template string with backticks, isn't it? Then Prettier should have an opinion here and should do the `` `${ spaced }` `` formatting.

aduth replied:
@jsnajdr That's what I would expect as well, but it does not appear to be working correctly.

jsnajdr replied:
It looks like a bug in `wp-prettier`. If the template literal is a styled component, with the `styled.span` tag, it's treated specially. Thanks for finding and reporting!

jsnajdr replied:
I created a `wp-prettier` fix in this commit: https://github.com/Automattic/wp-prettier/commit/01fd780c7c5a8edd9e3ef7daea49e650fa42cee5

A fix was needed for all the template literals that have a recognized format (CSS, GraphQL, HTML...) and where the string itself is being formatted.

I usually release such patches when a new upstream version comes out. In this case, I'll be waiting for `prettier@1.19.2`.

ItsJonQ commented on packages/components/src/range-control/README.md:
```diff
@@ -160,13 +168,41 @@ If no value exists this prop contains the slider starting position.
+const marks = [
+	{
+		value: 0,
+		label: '0',
```
I'm not sure if it's common yet :).
I can see cases where you may want to render `'10%'` instead
